File: recipes/llm-voice-assistant/android/llm-voice-assistant/src/main/java/ai/picovoice/llmvoiceassistant/TPSProfiler.java
package ai.picovoice.llmvoiceassistant;

public class TPSProfiler {
    private int numTokens;
    private long startSec;
    private long endSec;

    public TPSProfiler() {
        this.numTokens = 0;
        this.startSec = 0;
        this.endSec = 0;
    }

    public void tock() {
        if (this.startSec == 0) {
            this.startSec = System.nanoTime();
        } else {
            this.endSec = System.nanoTime();
            this.numTokens += 1;
        }
    }

    public double tps() {
        double tps = this.numTokens / ((this.endSec - this.startSec) / 1e9);
        this.numTokens = 0;
        this.startSec = 0;
        this.endSec = 0;
        return tps;
    }
}

File: recipes/llm-voice-assistant/ios/LLMVoiceAssistantDemo/TPSProfiler.swift
import Foundation

class TPSProfiler {
    private var numTokens: Int
    private var startSec: UInt64
    private var endSec: UInt64

    init() {
        self.numTokens = 0
        self.startSec = 0
        self.endSec = 0
    }

    func tock() {
        if self.startSec == 0 {
            self.startSec = DispatchTime.now().uptimeNanoseconds
        } else {
            self.endSec = DispatchTime.now().uptimeNanoseconds
            self.numTokens += 1
        }
    }

    func tps() -> Double {
        let elapsedSec = Double(self.endSec - self.startSec) / 1e9
        let tps = elapsedSec > 0 ? Double(self.numTokens) / elapsedSec : 0.0

        self.numTokens = 0
        self.startSec = 0
        self.endSec = 0

        return tps
    }
}
